### ISSUE TITLE
Overriding options for Dropdown field triggering validation

### ISSUE THREAD

**Issue by jimirobaer** [verbb/formie#1137]
### Describe the bug

We've had discussions in the past about how overriding the options of dropdown fields (or any other fields in Formie where options are set) wasn't really possible. Since 1.5, you have made changes on how this could work (using snapshots, referencing issue #415 and #578).

However, I am still not really sure on how to properly override options for these fields without triggering the validation.

If I create a new dropdown inside my form and fill in one (or many) option(s) (as required, because I can't create a dropdown field without options), and override using the example code from [the docs](https://verbb.io/craft-plugins/formie/docs/template-guides/rendering-fields#override-field-settings), I get a validation issue on submit (in the JSON output below).

There is no issue if I manually add two options in the dropdown with "override-one" and "override-two" as the value. But that defeats the purpose of being able to dynamically populate the the fields. If this is (still) the intended use, then I don't think the documentation is very clear about this limitation.

I tried the override on both custom templating and using the default `craft.formie.renderForm` function.

### Steps to reproduce

1. Create a dropdown field and entering a single (or multiple) options
2. Override the options referencing the field and adding the correct syntax
3. Submitting the form

### Form settings

- Multi-page form: No
- Submission Method: Ajax
- Client-side Validation: Yes
- Custom Form Templates: Yes

### Craft CMS version

4.2.5.2

### Plugin version

2.0.13

### Multi-site?

Yes

### Additional context

**Output from JSON after submit**

```
{
   "success":false,
   "submissionId":null,
   "currentPageId":134,
   "nextPageId":134,
   "nextPageIndex":0,
   "totalPages":1,
   "redirectUrl":"",
   "submitActionMessage":"Submission saved.",
   "events":[
      
   ],
   "errors":{
      "overrideDropdown":[
         "Subject is invalid."
      ]
   },
   "pageFieldErrors":{
      "134":{
         "overrideDropdown":[
            "Subject is invalid."
         ]
      }
   },
   "errorMessage":"Couldn’t save submission due to errors."
}
```

**Field settings in Formie**

<img width="1211" alt="Screenshot 2022-10-24 at 16 48 36" src="https://user-images.githubusercontent.com/8984769/197555659-28cc5e5d-7ac5-45c2-8e1c-b7d2b44d108a.png">

**Comment by engram-design** [comment]
Looks like this is a regression, as that should totally be possible through the use of snapshots. 

However, looks like this is only an issue of `http` domains. Can you confirm if you're using HTTPS or just HTTP? That's the only way I was able to reproduce this, and the Cookie settings for Craft don't seem to be taking effect for non-SSL.

**Comment by jimirobaer** [comment]
> However, looks like this is only an issue of `http` domains. Can you confirm if you're using HTTPS or just HTTP? That's the only way I was able to reproduce this, and the Cookie settings for Craft don't seem to be taking effect for non-SSL.

I tested it locally on an HTTP domain but can reproduce the same issue running it on a server with the domain on HTTPS. Attached is the JSON response, identical to the one above.

```
{
   "success":false,
   "submissionId":null,
   "currentPageId":769,
   "nextPageId":769,
   "nextPageIndex":0,
   "totalPages":1,
   "redirectUrl":"",
   "submitActionMessage":"Submission saved.",
   "events":[
      
   ],
   "errors":{
      "overrideRegion":[
         "Regio is ongeldig."
      ]
   },
   "pageFieldErrors":{
      "769":{
         "overrideRegion":[
            "Regio is ongeldig."
         ]
      }
   },
   "errorMessage":"Couldn’t save submission due to errors."
}
```

**Comment by engram-design** [comment]
Just to confirm what templating you're using:

```twig
{% do form.setFieldSettings('overrideRegion', {
    options: [
        { label: 'Override One', value: 'override-one', isDefault: false },
        { label: 'Override Two', value: 'override-two', isDefault: false },
    ],
}) %}
```

As that seems to work for me!

**Comment by jimirobaer** [comment]
> Just to confirm what templating you're using:
> 
> ```
> {% do form.setFieldSettings('overrideRegion', {
>     options: [
>         { label: 'Override One', value: 'override-one', isDefault: false },
>         { label: 'Override Two', value: 'override-two', isDefault: false },
>     ],
> }) %}
> ```
> 
> As that seems to work for me!

Just to make sure I'm doing everything the way it's supposed to:

1. I installed a clean installation of Craft CMS with Formie as the sole plugin. I am running DDEV with local certificate and the site on HTTPS;
2. Created an empty form (with no custom template), added a single Dropdown field ("dropdownField") and populated it with one option, set as Default; 
3. Added the following template code in index.twig:
```
{% set form = craft.formie.forms.handle('contactForm').one() %}

{% do form.setFieldSettings('dropdownField', {
    options: [
        { label: 'Override One', value: 'override-one', isDefault: false },
        { label: 'Override Two', value: 'override-two', isDefault: false },
    ],
}) %}

{{ craft.formie.renderForm(form) }}
```

And I still get an issue after submitting the dropdown, identical to the ones above;

**Settings**
<img width="1229" alt="Screenshot 2022-10-25 at 12 10 48" src="https://user-images.githubusercontent.com/8984769/197746525-fb09449c-e65f-48a3-9031-2b593f407176.png">

**Submission**
<img width="1431" alt="Screenshot 2022-10-25 at 12 14 51" src="https://user-images.githubusercontent.com/8984769/197747377-030f278e-a8bb-4c07-97d6-4583ca83710f.png">

**Comment by engram-design** [comment]
That all looks completely correct to me. I've even spun up a brand new site just to triple test this but still working for me.

Anything worth noting about your install - even for your simple reproduction? Special config (like swapping the cache or session stored to something else like Redis)?

Otherwise will have to keep at trying to replicate.

**Comment by jimirobaer** [comment]
Not at all, tried on two different local servers (Nginx/Apache) and identical issues when deployed on a server.

Would it help if I sent you database and project with composer.json?

**Comment by engram-design** [comment]
Okay, I _think_ I've figured out what's going on here! Should be fixed for the next release. To get this early, change your `verbb/formie` requirement in `composer.json` to:

```
"require": {
  "verbb/formie": "dev-craft-4 as 2.0.13",
  "...": "..."
}
```

Then run `composer update`.

**Comment by jimirobaer** [comment]
Cool! That seemed to do the trick. Great, thanks for the quick reply and solution as usual.

**Comment by engram-design** [comment]
Fixed in [2.0.14](https://github.com/verbb/formie/releases/tag/2.0.14)

**Comment by jimirobaer** [comment]
Sorry for having to reopen this issue, but I note that I am experiencing the initial problem again when I apply Blitz to the page. I was able to reproduce it on the same 'vanilla' environment as the initial bug report. 

I supplied all the necessary JavaScript to refresh the tokens (both the CSRF and captchas) as described on the helpful article on [your website](https://verbb.io/craft-plugins/formie/docs/template-guides/cached-forms). I can verify that the tokens indeed get refreshed on page refresh.

I can only describe the issue like this: it seems that when the page (containing the form) is first cached, submitting the form (selecting "Override One") works. However, if I refresh the page and try to submit the form again, I get the same error message from my initial post. It does not matter which option I select.

I can reproduce the issue both using Page Reload and AJAX submission methods.

I have updated Formie to 2.0.16 and Craft CMS to 4.3.1.

https://user-images.githubusercontent.com/8984769/201060862-fb547d4a-4313-428a-96f8-75ccc7638a33.mp4

**Comment by engram-design** [comment]
If you're using Blitz, that would make sense that the settings revert back. That's because with Blitz, you're not running any Twig on the page, which `form.setFieldSettings()` is. There's no way to modify anything about the form when using Blitz, because it relies on running Twig, or access to server-side code to apply field settings. Having these available through JS would be a security issue.

Even if we were to employ a "session" of some sort that upon the Twig code running once, it "saves" your values against the field, there's no way of knowing when to turn this off.

The only way around this would be to either disable Blitz for the page, look at if it's possible to have dynamic sections of the page with Blitz still enabled, or use a Hidden field and style it like a dropdown instead.

**Comment by cvitko** [comment]
I am having the same issue as jimirobaer described initially. Overriding dropdown options and the validation says the field is invalid. 
I am not using Blitz. 
I am on the last version 2.0.17. 
Can you please check if the fix from 2.0.14 is still working. Thanks.

**Comment by engram-design** [comment]
Fixed for the next release. To get this early, change your `verbb/formie` requirement in `composer.json` to:

```
"require": {
  "verbb/formie": "dev-craft-4 as 2.0.17",
  "...": "..."
}
```

Then run `composer update`.

**Comment by cvitko** [comment]
I updated but still not working, same behaviour.

**Comment by engram-design** [comment]
@cvitko Can you share anything more about your setup, as it's working for me? Page-reload or Ajax? Custom rendering or `renderForm()`?

One thing I will mention is that when using Ajax forms, once submitted the dynamic settings will be reset, so if you try and send the form again straight away without a page refresh, you'll likely hit this error. This is because we need a way to remove the dynamic settings, against the field.

**Comment by cvitko** [comment]
@engram-design So I had to do a bit of experimenting. I created new testing form and it is really working. But when I add for example a text field and set visibility to hidden, after submitting the dropdown field is invalid.
```
{% set test = craft.formie.forms({ handle: 'test' }).one() %}

{% do test.setFieldSettings('dropdown', {
    options: [
        {label: 'Hello World', value: 'hello-world', isDefault: false},
        {label: 'Example', value: 'example', isDefault: false},
    ]
}) %}

{% do test.setFieldSettings('text', {
    visibility: 'hidden'
}) %}

{{ craft.formie.renderForm(test) }}
```

**Comment by engram-design** [comment]
Ah yes, good point. Fixed for the next release. To get this early, change your `verbb/formie` requirement in `composer.json` to:

```
"require": {
  "verbb/formie": "dev-craft-4 as 2.0.18",
  "...": "..."
}
```

Then run `composer update`.

**Comment by engram-design** [comment]
Fixed in [2.0.19](https://github.com/verbb/formie/releases/tag/2.0.19)
